src/System.Management.Automation/engine/Utils.cs

```diff
@@ -33,6 +34,223 @@ namespace System.Management.Automation
+        internal static BigInteger ParseBinary(ReadOnlySpan<char> digits, bool unsigned)
+        {
+            if (!unsigned)
+            {
+                if (digits[0] == '0')
+                {
+                    unsigned = true;
+                }
+                else
+                {
+                    switch (digits.Length)
+                    {
+                        // Only accept sign bits at these lengths:
+                        case 8: // byte
+                        case 16: // short
+                        case 32: // int
+                        case 64: // long
+                        case 96: // decimal
+                        case int n when n >= 128: // BigInteger
+                            break;
+                        default:
+                            // If we do not flag these as unsigned, bigint assumes a sign bit for any (8 * n) string length
+                            unsigned = true;
+                            break;
+                    }
+                }
+            }
+
+            // Only use heap allocation for very large numbers
+            const int MaxStackAllocation = 512;
+
+            // Calculate number of 8-bit bytes needed to hold the input,  rounded up to next whole number.
+            int outputByteCount = (digits.Length + 7) / 8;
+            int outputByteIndex = outputByteCount - 1;
+
+            Span<byte> outputBytes = outputByteCount <= MaxStackAllocation ? stackalloc byte[outputByteCount] : new byte[outputByteCount];
+
+            // We need to be prepared for any partial leading bytes, (e.g., 010|00000011|00101100), or cases
+            // where we only have less than 8 bits to work with from the beginning.
+            //
+            // Walk bytes right to left, stepping one whole byte at a time (if there are any whole bytes).
+            int byteWalker;
+            for (byteWalker = digits.Length - 1; byteWalker >= 7; byteWalker -= 8)
+            {
+                // Use bit shifts and binary-or to sum the values in each byte.  These calculations will
+                // create values higher than a single byte, but the higher bits will be stripped out when cast
+                // to byte.
+                //
+                // The low bits are added in separately to allow us to strip the higher 'noise' bits before we
+                // sum the values using binary-or.
+                //
+                // N.B.: This code has been tested against a straight for loop iterating through the byte, and in no
+                // circumstance was it faster or more effective than this unrolled version.
+                outputBytes[outputByteIndex--] =
+                    (byte)(((digits[byteWalker - 7] << 7)
+                        | (digits[byteWalker - 6] << 6)
+                        | (digits[byteWalker - 5] << 5)
+                        | (digits[byteWalker - 4] << 4))
+                        | (((digits[byteWalker - 3] << 3)
```

Comment:
The inputs are char `'0'` and `'1'`, and their bit patterns are `00110000` and `00110001`. 

With 8 input chars, we need 

00110001
00110001
00110001
00110001
00110001
00110001
00110001 We only need the far right bit out of each char, which we bit-shift with `<<`. we can do less work by processing bits

Reply:
Why not `( (7)|(6)|(5)|(4)|(3)|(2)|(1)|(0) ) & 0b1111`?

Reply:
Thanks for clarification!

I think we need reformat the code and add the comment.

Reply:
I would appreciate some suggestions on how to approach the reformatting. The actual behaviour of the code *is* documented in the comments (although perhaps that can also be improved upon).

I suspect it will be necessary to skirt StyleCop rules heavily to make this much more readable than it is?

Reply:
1. We could add the well-formatted sample `(    (7)|(6)|(5)|(4)    )    |    (    ( (3)|(2)|(1)|(0) ) & 0b1111    )` in your comment.
2. We could add white spaces in code for better readability.

Reply:
Sounds good!

I think this *might* annoy StyleCop, but... such is the price of readability. 😄